### ISSUE TITLE
cri-o hangs; PLEG is not healthy

### ISSUE THREAD

**Issue by mariusgrigoriu** [cri-o/cri-o#2584]
**Description**
We found one of our k8s nodes in a NotReady state. The reason is that PLEG is not healthy. Upon closer inspection we found that some crictl commands are timing out after waiting long.

**Steps to reproduce the issue:**
Unknown. Happens intermittently.

**Describe the results you received:**
- Kubelet reports PLEG is unhealthy.
- crictl ps works fine
- crictl pods works fine
- crictl images hangs
- crictl inspect also hangs
- crictl inspectp works fine

Restarting cri-o daemon solves the problem until some point in the future when it repeats.

**Describe the results you expected:**
cri-o daemon remains responsive

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `crio --version`:**

```
crio version 1.14.5
commit: "1b510ce8cea807d17215244dca46d2fa2b1ef71f"
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Core OS 2079.6.0
Kubernetes 1.14.3

**Comment by SleepyBrett** [comment]
After one of the nodes went into this state I went ahead and hit it with a `kill -6` the abort portion of the logs is in this gist: https://gist.github.com/SleepyBrett/0b4f5bc083c5b6466710f457290f9784

**Comment by matthias50** [comment]
I ran into a problem with very similar symptoms with docker and it turned out to be an overflow issue in systemd. See https://github.com/systemd/systemd/issues/11809. I was able to work around it by having a cron job which does ```systemctl daemon-reload``` daily.

**Comment by mariusgrigoriu** [comment]
What's interesting is that call to list images was freezing, so we expect to be no connection to the cgroup driver, but who knows. We're using cgroupfs so will be trying #2601 and reporting back.

**Comment by mariusgrigoriu** [comment]
We're still seeing this happen using commit `906b95a` from the 1.14 release branch. How might we track this down?

**Comment by mrunalp** [comment]
@vrothberg @mtrmac @haircommander We updated a lot of code for images after 1.14.4. We are also hitting this issue in CI. ptal.

**Comment by haircommander** [comment]
Where are we seeing it in CI?

**Comment by mrunalp** [comment]
@haircommander in the wider OpenShift e2e.

**Comment by mrunalp** [comment]
Also adding @nalind since @mtrmac suspects it could be storage locks.

**Comment by mtrmac** [comment]
@nalind
```
goroutine 1320454 [semacquire, 11 minutes]:
github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*lockfile).lock(0xc000948140, 0x1000000)
github.com/cri-o/cri-o/server.(*Server).ContainerStatus(0xc00043d680, 0x1dffea0, 0xc005a98690, 0xc003b582c0, 0xc000b9aa90, 0x0, 0x0)

goroutine 1345323 [semacquire, 1 minutes]:
github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*lockfile).lock(0xc000948140, 0x1000000)
github.com/cri-o/cri-o/pkg/storage.(*imageService).ListImages(0xc00096e180, 0xc000950000, 0x0, 0x0, 0x1, 0x1, 0x1ae9000, 0x1e069e0, 0x0)

goroutine 1345301 [semacquire, 1 minutes]:
github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*lockfile).lock(0xc000948140, 0x1000000)
github.com/cri-o/cri-o/server.(*Server).ContainerStatus(0xc00043d680, 0x1dffea0, 0xc0037b0510, 0xc000a80100, 0xc000b40020, 0x0, 0x0)

goroutine 1317711 [semacquire, 12 minutes]:
github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*lockfile).lock(0xc000948140, 0x1000000)
github.com/cri-o/cri-o/server.(*Server).ContainerStatus(0xc00043d680, 0x1dffea0, 0xc000898060, 0xc004796200, 0xc000a78630, 0x0, 0x0)

…
```
There are quite a few coroutines blocked on `github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*imageStore).RLock`.

In the meantime, there is
```
goroutine 1317778 [runnable]:
github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*layerStore).Delete(0xc00086af70, 0xc000dafa80, 0x40, 0xc003810e01, 0x0)
github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*store).DeleteContainer(0xc000537c20, 0xc0027b8000, 0x40, 0x0, 0x0)
github.com/cri-o/cri-o/server.(*Server).RemoveContainer(0xc00043d680, 0x1dffea0, 0xc0049c23c0, 0xc000b9a660, 0x0, 0x0, 0x0)
```
probably (because it got to `layerStore.Delete`) holding the `imageStore` exclusive lock.

(and also
```
goroutine 1265698 [semacquire, 12 minutes]:
github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*lockfile).Lock(0xc000948050)
github.com/cri-o/cri-o/vendor/github.com/containers/storage.(*layerStore).Lock(0xc00086af70)
github.com/cri-o/cri-o/server.(*Server).PullImage(0xc00043d680, 0x1dffea0, 0xc00310c660, 0xc00153e9a0, 0x0, 0x1db1840, 0xc000100e40)
```
, FWIW, but AFAICS harmless)

So, `DeleteContainer` seems to be holding the image lock while `layerStore.Delete` is taking probably most of those  12 minutes. Is it possible that it just takes that much to delete the image? If not, I can’t quite see how there could be a loop, at best there is the
```go
for mountCount > 0
```
one but I can’t see how `mountCount` is not decreasing.

**Comment by nalind** [comment]
I wouldn't expect it to take that long to delete an image unless the system was unusually heavily loaded, as it boils down to unmounting the container's read-write layer (which has either succeeded or failed with ENOENT by the time we get to line 811), and then removing the layer.

The backtrace hasn't yet reached the point where it tries to remove the layer from storage, so thus far it's only called `umount()`, and is partway through the process of updating its record of which layers are mounted, and where the mounted ones are mounted.  I don't see a path for a loop to get out of control at `for mountCount > 0`, either.  When we have a node in this state, do we have a way to attach and start observing the goroutine?

**Comment by saschagrunert** [comment]
@haircommander do you think we can close this issue? It looks like that it does not happen any more with CRI-O versions > 1.15, right?

**Comment by haircommander** [comment]
yup, and cri-o 1.14.10 also has the fix

**Comment by skraga** [comment]
Hi,
we're experiencing this or very similar issue on a special node group where we run a batch (temporary) workload. Pods are being continuously created/deleted on them.
crio: version 1.15.2-1.el7
kernel: 3.10.0-1062.12.1.el7.x86_64
OS: CentOS Linux release 7.7.1908 (Core)

It causes these nodes to continuously flap (Ready/NotReady) by PLEG.

**Comment by haircommander** [comment]
@skraga I am unsure this should go with this issue (as this is resolved in 1.15.2 I think)

However, can you try upgrading to 1.15.3? If it still happens, can you open a new issue to track?
